lom/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -115,63 +166,449 @@ func (plmgr *PluginManager) getPluginMetadata(plgname string) (plugins_common.IP
+    wg.Wait()
+
+    // TODO: Goutham/Renuka : verify if needed, Deregister the client with server
+    plmgr.clientTx.DeregisterClient()
+
+    lomcommon.DeregisterForSysShutdown(APP_NAME_DEAULT + ProcID)
+    // calls blocks untill all the registered clients to be deregistered for shutdown
+    lomcommon.DoSysShutdown(0) /* TODO: Goutham : Is it Needed ?? with '0' we are waiting indefinitly untill all
+       the registered clients to be deregistered for shutdown */
+
+    // TODO: Goutham/Renuka : verify if needed, Send shutdown response to server
+    //plmgr.sendResponseToEngine(&lomipc.MsgEmptyResp{})
+
+    // Waiting for all goroutines created by goroutinetracker untill GOROUTINE_CLEANUP_TIMEOUT_DEFAULT(30 sec default) to finish
+    // TODO: Goutham :  Do we need to log panic if unable to get shutdown from plugin in neded time?
+    if !lomcommon.GetGoroutineTracker().WaitAll(GOROUTINE_CLEANUP_TIMEOUT_DEFAULT) {
+        lomcommon.LogInfo("In handleShutdown(): Timed out waiting for goroutines to finish")
+
+        // Print running goroutine still left info to syslog
+        lomcommon.PrintGoroutineInfo("", true)
+    }
+
+    // Clear the plugin and pluginMetadata maps
+    plmgr.plugins = nil
+    plmgr.pluginMetadata = nil
+
+    // Exit the process
+    lomcommon.LogInfo("In handleShutdown(): Exiting process")
+    os.Exit(0)
+
+    return nil
+}
+
+/*
+ * Shutdown the plugin.
+ * Invoke the plugin's shutdown method
+ * Starts shutdown grace timer at its end. If the timer expires or if the shutdown is complete before timer expiry,
+ * then shutdown is treated as complete.
+ */
+func (plmgr *PluginManager) shutdownPlugin(pluginname string) {
+    plugin, ok := plmgr.plugins[pluginname]
+    if !ok {
+        lomcommon.LogInfo("In shutdownPlugin(): Plugin %s not found", pluginname)
+        return
+    }
+
+    if plugin == nil {
+        lomcommon.LogInfo("In shutdownPlugin(): Plugin %s is nil", pluginname)
+        return
+    }
+
+    lomcommon.LogInfo("In shutdownPlugin(): Shutting down plugin %s", pluginname)
+
+    pluginmetadata, ok := plmgr.getPluginMetadata(pluginname)
+    if !ok {
+        lomcommon.LogError("handleRequest() : Plugin %s metadata not initialized", plugin)
+        return
+    }
+
+    if pluginmetadata == nil {
+        lomcommon.LogError("handleRequest() : Plugin %s metadata is nil", plugin)
+        return
+    }
+
+    // This is to handle a case for where plugin may be shutdown via misbehaving specific plugin logic
+    if pluginmetadata.GetPluginStage() == plugins_common.PluginStageDisabled {
+        lomcommon.LogInfo("In shutdownPlugin(): Plugin is already disabled %s", pluginname)
+        return
+    }
+
+    shutdownRespCh := make(chan error, 1)
+
+    lomcommon.GetGoroutineTracker().Start("plg_mgr_shutdownPlugin_"+pluginname+"_"+lomcommon.GetUUID(),
+        func() {
+            pluginmetadata.SetPluginStage(plugins_common.PluginStageShutdownStarted)
+            retStatus := plugin.Shutdown()
```

Comment:
You don't want a misbehaving plugin to hijack. So have a reasonable timeout to stop waiting